Use correct API for fetching user lists

ReeceHoffmann opened the pull request:
Changes:
 - Converts `ManageUsers` and `UserList` to function based components
 - Changes `User` view to use `/administrators/users` api
 - Adds infinite scroll query for fetching users
 - Implements new react-query compatible scroll list

ReeceHoffmann committed:
fix: use correct API for fetching user lists

igboyes commented on src/js/users/components/List.tsx:
```diff
@@ -0,0 +1,48 @@
+
+const renderRow = (item: User) => (
+    <UserItem key={item.id} id={item.id} handle={item.handle} administrator_role={item.administrator_role} />
+);
```
`function`

igboyes commented on src/js/administration/hooks.ts:
```diff
@@ -10,3 +14,11 @@ function fetchSettings() {
+        hasPermission: account ? hasSufficientAdminRole(requiredRole, account.administrator_role) : null,
+        isLoading,
+    };
+}
```
Let's start documenting. 

https://jsdoc.app/

igboyes commented on src/js/base/ScrollList.tsx:
```diff
@@ -59,3 +60,44 @@ export const ScrollList = ({ page, documents, pageCount, onLoadNextPage, renderR
+    isLoading: boolean;
+    items: unknown[];
+    renderRow: (item: unknown) => void;
+};
```
```suggestion
};

```

igboyes commented on src/js/base/ScrollList.tsx:
```diff
@@ -59,3 +60,44 @@ export const ScrollList = ({ page, documents, pageCount, onLoadNextPage, renderR
+            {loading}
+        </StyledScrollList>
+    );
+};
```
```suggestion

    return (
        <StyledScrollList>
            {entries}
            {isLoading && <LoadingPlaceholder margin="20px" />}
        </StyledScrollList>
    );
};
```

igboyes commented on src/js/users/components/Item.tsx:
```diff
@@ -59,7 +57,11 @@ const EditButton = ({ id }) => (
 );
 
-export const UserItem = ({ id, handle, administrator_role, canEdit }) => {
+export const UserItem = ({ id, handle, administrator_role }) => {
```
`function`

igboyes commented on src/js/users/components/List.tsx:
```diff
@@ -0,0 +1,48 @@
+ * Ian Boyes
+ *
+ * @exports UsersList
+ */
```
This is pretty useless. I think we should only have module comments if they have actual sentences in them.

igboyes commented on src/js/users/components/List.tsx:
```diff
@@ -0,0 +1,48 @@
+type UsersListProps = {
+    term: string;
+};
+export function UsersList({ term }: UsersListProps) {
```
```suggestion
};

export function UsersList({ term }: UsersListProps) {
```

igboyes commented on src/js/users/components/Users.tsx:
```diff
@@ -0,0 +1,42 @@
+            <CreateUser />
+        </>
+    );
+}
```
Swap these to avoid `!`.

igboyes commented on src/tests/fake/user.tsx:
```diff
@@ -63,7 +63,7 @@ type Query = {
 };
 
-export function mockGetUsersAPI(users: Array<User>, query?: Query) {
+export function mockFindUsersAPI(users: Array<User>, query?: Query) {
```
```suggestion
export function mockFindUsersApi(users: Array<User>, query?: Query) {
```

igboyes commented on src/js/administration/hooks.ts:
```diff
@@ -10,3 +14,11 @@ function fetchSettings() {
 }
+
+export function useCheckAdminRole(requiredRole: AdministratorRoles) {
+    const { data: account, isLoading }: { data: Account; isLoading: boolean } = useGetAccount();
```
Isn't typing inferred here or is return type for `useGetAccount` not defined?

ReeceHoffmann replied:
This was done since the default typing being returned for `useQuery` had the datatype as `unknown`. That said the reactQueryAPI provides for a way to specify the returned data type. Next push will update the queries typing so that types can be automatically inferred for all usages of the request

igboyes commented on src/js/administration/components/administrators/__tests__/Create.test.tsx:
```diff
@@ -86,7 +86,7 @@ describe("<CreateAdministrator>", () => {
         expect(await screen.findByRole("option", { name: users[1].handle })).toBeInTheDocument();
 
-        const get_users_scope = mockGetUsersAPI([users[0]], {
+        const get_users_scope = mockFindUsersAPI([users[0]], {
```
What does 'scope' mean here?

```suggestion
        const getUsersScope = mockFindUsersAPI([users[0]], {
```

ReeceHoffmann replied:
`scope` is in reference to the default `Scope` class instance returned by nock when mocking an API route. Honestly not super pumped by the name myself, but the goal was to keep it obvious that the variable is an instance of `Scope` while also specifying which request it is mocking

igboyes replied:
That makes sense.

ReeceHoffmann committed:
chore: address review

  - adds documentation for administrator queries

ReeceHoffmann committed:
Merge branch 'main' into reecehoffmann/ui-47-use-new-admin-api-in-admin-views

ReeceHoffmann committed:
fix: restore tests post merge

ReeceHoffmann committed:
chore: improve documentation

 - Add documentation for many exported functions
 - Update some existing documentation from JSDocs to TSDocs
 - Align many component file names with underlying component name
 - improve UsersList mapping readability
 - Restore settings fetching via sagas

igboyes commented on src/js/account/api.js:
```diff
@@ -109,3 +109,13 @@ export const logout = () => Request.get("/account/logout");
+ *
+ * @returns A Promise resolving to the account data
+ */
+export const FetchAccount = () =>
```
```suggestion
export function fetchAccount () {
```

igboyes commented on src/js/administration/querys.ts:
```diff
@@ -1,42 +1,97 @@
+/**
+ * Factory object for generating settings query keys
+ */
+export const settingsKeys = {
```
```suggestion
export const settingsQueryKeys = {
```

igboyes commented on src/js/administration/querys.ts:
```diff
@@ -1,42 +1,97 @@
+/**
+ * Factory object for generating user query keys
+ */
 export const userKeys = {
```
```suggestion
export const userQueryKeys = {
```

igboyes commented on src/js/administration/sagas.js:
```diff
@@ -1,19 +1,19 @@
 import { GET_SETTINGS, UPDATE_SETTINGS } from "../app/actionTypes";
 import { apiCall } from "../utils/sagas";
-import * as settingsAPI from "./api";
+import { legacyFetchSettings as ApiFetchSettings, updateSettings as ApiUpdateSettings } from "./api";
```
```suggestion
import { legacyFetchSettings as apiFetchSettings, updateSettings as apiUpdateSettings } from "./api";
```

igboyes commented on src/js/administration/types.ts:
```diff
@@ -1,13 +1,24 @@
+    BASE = "base",
 }
+
+export type permissionQueryResult = {
```
I believe this is conventional.
```suggestion
export type PermissionQueryResult = {
```

igboyes commented on src/js/users/components/UserItem.tsx:
```diff
@@ -45,21 +43,38 @@ const AdminTagContainer = styled(UserContainer)`
-        </Label>
-    </AdminTagContainer>
-);
+function AdminTag({ administrator_role }) {
```
```suggestion
function AdminTag({ administratorRole }) {
```

igboyes commented on src/js/users/components/__tests__/ManageUsers.test.jsx:
```diff
@@ -0,0 +1,80 @@
+import { AdministratorRoles } from "../../../administration/types";
+import { ManageUsers } from "../ManageUsers";
+
+const createReducer = (state, history) =>
```
```suggestion
function createReducer (state, history) {
```

ReeceHoffmann committed:
docs: update account API docs to TSDoc

ReeceHoffmann committed:
Merge branch 'main' into reecehoffmann/ui-47-use-new-admin-api-in-admin-views

ReeceHoffmann committed:
refactor: increase naming consistency

ReeceHoffmann committed:
refactor: reduce use of snakeCase variable names

ReeceHoffmann committed:
fix: prevent snapshots failures from changing dates

igboyes commented:
What is going on with this @ReeceHoffmann?

ReeceHoffmann commented:
> What is going on with this @ReeceHoffmann?

Should have been good to go. Looks like CI failures due to the clone reference test not accounting for changes to the real world date. Adjusted to always have the passed date be exactly one year in the past

igboyes commented on src/js/account/api.ts:
```diff
@@ -0,0 +1,160 @@
+/**
+ * Defines methods used for calling account API endpoints.
```
They aren't methods.

igboyes commented on src/js/administration/api.ts:
```diff
@@ -0,0 +1,66 @@
+import { Request } from "../app/request";
+/**
```
Add module docstring if you are going to do it elsewhere.
```suggestion
import { Request } from "../app/request";

/**
```

igboyes commented on src/js/administration/api.ts:
```diff
@@ -0,0 +1,66 @@
+ */
+export function fetchAdministratorRoles() {
+    return Request.get("/admin/roles").then(response => response.body);
+}
```
You are missing return types here, but you included them in other `api.ts` modules.

igboyes commented on src/js/administration/utils.ts:
```diff
@@ -9,10 +12,20 @@ enum AdministratorPermissionsLevel {
+ * @param userRole - The administrator role of the user
+ */
+
 export const hasSufficientAdminRole = (requiredRole: AdministratorRoles, userRole: AdministratorRoles): boolean => {
```
Change to `function`.

igboyes commented on src/js/base/ScrollList.tsx:
```diff
@@ -59,3 +60,39 @@ export const ScrollList = ({ page, documents, pageCount, onLoadNextPage, renderR
+    renderRow: (item: unknown) => void;
+};
+
+export const StreamlinedScrollList = ({
```
* What is streamlined compared to `ScrollList`? Need a docstring.
* Use `function`.

ReeceHoffmann replied:
The key difference was that `StreamlinedScrollList` was designed around using `react-queries` infinite scroll queries. Ended up taking the approach of instead renaming `StreamlinedScrollList` to `ScrollList` and the original `ScrollList` to `LegacyScrollList` as usage of the older version should be phased out as more redux logic gets stripped out. (or as we move to pagination)

igboyes commented on src/js/users/components/UserItem.tsx:
```diff
@@ -45,21 +43,38 @@ const AdminTagContainer = styled(UserContainer)`
+ * @returns A single user element
+ */
+
+export function UserItem({ id, handle, administratorRole }) {
```
Typing?

igboyes commented on src/js/users/components/ManageUsers.tsx:
```diff
@@ -0,0 +1,49 @@
+import { UsersList } from "./UsersList";
+
+/**
+ * The base user management view
```
What is it a base for?

ReeceHoffmann committed:
docs: update typing and documentation

ReeceHoffmann committed:
fix: update snapshots for `ScrollList` containing components

ghost commented:
:tada: This PR is included in version 3.13.1 :tada:

The release is available on [GitHub release](https://github.com/virtool/virtool-ui/releases/tag/3.13.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
